File: Client_Java/Client_Java/clientgui/gameGUI/GameGUI.java
package Client_Java.clientgui.gameGUI;

import Client_Java.WordyApp.WordyGame;
import Client_Java.clientgui.mainGUI.HomePageGUI;

import javax.swing.*;
import javax.swing.text.AttributeSet;
import javax.swing.text.BadLocationException;
import javax.swing.text.PlainDocument;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.*;
import java.util.Timer;
import java.util.TimerTask;


public class GameGUI extends JFrame {
    private ImageIcon bg;
    private JLabel round, user, time, score, word, sec, validation = new JLabel();
    private JButton shuffleButton;
    private JTextField ans;
    private Timer timer;
    private static WordyGame wordObj;
    private static String username, roomName, firstLetterLine, secondLetterLine;
    public GameGUI(WordyGame obj, String username, String roomName, Point guiLocation) {
        wordObj = obj;
        GameGUI.username = username;
        GameGUI.roomName = roomName;
        setBackground(Color.gray);

        String[] roomInfo = wordObj.getGameRoomInformation(roomName, username);
        String points = roomInfo[1];

        // Set the size and location of the GUI
        setSize(800, 650);
        setLocation(guiLocation);

        timer = new java.util.Timer();
        timer.scheduleAtFixedRate(new CountdownTask(), 0, 1000);

        //set frame title to be room name
        setTitle(roomName);

        // Set the background color of the JFrame
        bg = new ImageIcon("CORBA/src/res/login.jpg");
        setContentPane(new JLabel(bg));

        round = new JLabel("Round ");
        round.setFont(new Font("Times New Roman", Font.BOLD, 40));

        user = new JLabel(username);
        user.setFont(new Font("Times New Roman", Font.BOLD, 23));

        time = new JLabel("Timer");
        time.setFont(new Font("Times New Roman", Font.BOLD, 30));

        sec = new JLabel("10"); //edit the 10
        sec.setFont(new Font("Times New Roman", Font.BOLD, 30));


        score = new JLabel("Score: " + points);
        score.setFont(new Font("Times New Roman", Font.BOLD, 23));

        shuffleButton = new JButton("Shuffle");
        shuffleButton.setFont(new Font("Times New Roman", Font.BOLD, 17));
        shuffleButton.setBackground(Color.YELLOW);
        shuffleButton.setForeground(Color.BLUE);

        word = new JLabel("<html><div style='text-align:center;'>" + firstLetterLine +
                "<br><span style='display:inline-block;width:130px;text-align:center;'>" +
                secondLetterLine + "</span></div></html>");
        word.setFont(new Font("Times New Roman", Font.BOLD, 50));
        word.setVerticalAlignment(JLabel.TOP);

        String randomLetters = roomInfo[4];
        shuffleLetters(randomLetters);

        class UpperCaseDocument extends PlainDocument {  // CAPSLOCK
            @Override
            public void insertString(int offs, String str, AttributeSet a)
                    throws BadLocationException {
                if (str == null) {
                    return;
                }
                super.insertString(offs, str.toUpperCase(), a); // Convert text to upper case
            }
        }

        //DEVELOPER TOOL
        if(username.equals("admin")) {
            enableCheats(randomLetters);
        }

        ans = new JTextField();
        ans.setFont(new Font("Arial", Font.PLAIN, 20));
        ans.setDocument(new UpperCaseDocument()); // Set the custom document
        ans.setHorizontalAlignment(JTextField.CENTER);
        ans.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                String answer = ans.getText();

                if(!sec.getText().equals("0")) {
                    if (wordObj.submitWord(answer, username, Integer.parseInt(roomInfo[3]), roomName)) {
                        getContentPane().remove(validation);
                        validation = new JLabel("ANSWER IS VALID");
                        validation.setFont(new Font("Arial", Font.BOLD, 15));
                        validation.setForeground(Color.GREEN);
                        validation.setBounds(340, 500, 550, 45);
                        add(validation);
                    } else {
                        getContentPane().remove(validation);
                        validation = new JLabel("INVALID WORD FORMED");
                        validation.setFont(new Font("Arial", Font.BOLD, 15));
                        validation.setForeground(Color.RED);
                        validation.setBounds(300, 500, 550, 45);
                        add(validation);
                    }
                } else {
                    JOptionPane.showMessageDialog(null, "Sorry, the round time is up!", "Time is up!", JOptionPane.ERROR_MESSAGE);
                }

                ans.setText("");
                // Refresh the JFrame to show the JLabel
                getContentPane().validate();
                getContentPane().repaint();
            }
        }) ;

        shuffleButton.addActionListener(e -> shuffleLetters(randomLetters)) ;

        round.setBounds(320, -25, 350, 120);
        user.setBounds(10, -40, 200, 120);
        score.setBounds(685, -40, 200, 120);

        time.setBounds(357, 70, 200, 120);
        sec.setBounds(387, 105, 200, 120);


        word.setBounds(160, 230, 800, 120);

        ans.setBounds(120, 430, 550, 45);
        shuffleButton.setBounds(570, 395, 100, 25);

        add(round);
        add(user);
        add(time);
        add(score);
        add(sec);
        add(word);
        add(ans);
        add(shuffleButton);

        addWindowListener(new WindowAdapter() {
            @Override
            public void windowClosing(WindowEvent e) {
                wordObj.logout(username);
                wordObj.leaveGameRoom(username, roomName);
                System.exit(0);
            }
        });
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            wordObj.logout(username);
            wordObj.leaveGameRoom(username, roomName);
            System.exit(0);
        }));

        // Set the GUI to be visible
        setResizable(true);
        setVisible(true);
    }

    private class CountdownTask extends TimerTask {
        String previousTimeValue = "";
        int repetitionCount = 0;
        @Override
        public void run() {
            String[] roomInfo = wordObj.getGameRoomInformation(roomName, username);
            String currentTimeValue = roomInfo[0];

            //Checker to see if the timer repeats itself seven times in a row
            //Brings back to main menu if so
            if (currentTimeValue.equals(previousTimeValue)) {
                repetitionCount++;
                if (repetitionCount == 7) {
                    JOptionPane.showMessageDialog(null, "The game has unexpectedly ended! " +
                            " Returning to main menu.", "Game Room Ended", JOptionPane.ERROR_MESSAGE);
                    dispose();
                    new HomePageGUI(wordObj, username, getLocation());
                    timer.cancel();
                }
            } else {
                repetitionCount = 1; //Reset repetition count
            }
            previousTimeValue = currentTimeValue;

            sec.setText(roomInfo[0]);

            round.setText("Round " + roomInfo[3]);

            if (currentTimeValue.equals("0")) {
                timer.cancel();
                Point location = getLocation();
                if (!wordObj.endRound(username, roomName)) {
                    dispose();
                    new ScoresGUI(wordObj, username, roomName, location);
                } else {
                    dispose();
                    new ResultGUI(wordObj, roomName, username, location);
                    }
            }
        }
    }

    private void shuffleLetters(String randomLetters) {

        char[] lettersArray = randomLetters.toCharArray();

        //THE FISHER-YALES SHUFFLE ALGORITHM
        Random rand = new Random();
        for (int i = lettersArray.length - 1; i > 0; i--) {
            int index = rand.nextInt(i + 1);
            char temp = lettersArray[index];
            lettersArray[index] = lettersArray[i];
            lettersArray[i] = temp;
        }

        firstLetterLine = new String(lettersArray, 0, 10);
        secondLetterLine = new String(lettersArray, 10, 7);

        firstLetterLine = firstLetterLine.replaceAll("", " ").trim();
        secondLetterLine = secondLetterLine.replaceAll("", " ").trim();

        word.setText("<html><div style='text-align:center;'>" + firstLetterLine +
                "<br><span style='display:inline-block;width:130px;text-align:center;'>" +
                secondLetterLine + "</span></div></html>");
    }

    /*
    ONLY DEVELOPER ASSETS FROM HERE ON
    */
    private static String[] arrayOfValidWords;
    private static ArrayList<String> arrayOfCheatedAnswers;
    private static int cheatArrayIndex = 0,  printCheatWordsIndex = 0;
    private void enableCheats(String randomLetters) {

        JButton cheatModeButton = new JButton("HACK");
        cheatModeButton.setFont(new Font("Times New Roman", Font.BOLD, 17));
        cheatModeButton.setBackground(Color.RED);
        cheatModeButton.setForeground(Color.WHITE);
        cheatModeButton.setBounds(570, 365, 100, 25);
        add(cheatModeButton);
        /*
        cheat mode populates an array of correct answers and sets the ans field to the first array word.
        if user presses enter then it should populate the next word. press cheatmode once
         */
        cheatModeButton.addActionListener(e -> {
            scanTextFile();
            arrayOfCheatedAnswers = new ArrayList<>();
            for (String word : arrayOfValidWords) {
                boolean canFormWord = canFormWordFromLetters(word, randomLetters);
                if (canFormWord) {
                    if((printCheatWordsIndex % 20 - 1) == 0) {
                        System.out.print(word);
                    } else if (printCheatWordsIndex % 20 == 0) {
                        System.out.println(word + ", ");
                    } else {
                        System.out.print(word + ", ");
                    }
                    printCheatWordsIndex++;
                    arrayOfCheatedAnswers.add(word);
                }
            }
            showHackedAnswers();
            getContentPane().validate();
            getContentPane().repaint();
        });
    }
    private void scanTextFile() {
        ArrayList<String> listOfStrings = new ArrayList<>();

        try (BufferedReader bf = new BufferedReader(new FileReader("CORBA/src/res/words.txt"))) {
            String line = bf.readLine();

            while (line != null) {
                listOfStrings.add(line);
                line = bf.readLine();
            }
        } catch (IOException e) {
            JOptionPane.showMessageDialog(this, "Admin files not found, cannot activate cheats", "Error", JOptionPane.ERROR_MESSAGE);
            e.printStackTrace();
        }

        arrayOfValidWords = listOfStrings.toArray(new String[0]);
    }
    private void showHackedAnswers() {
        if (cheatArrayIndex < arrayOfCheatedAnswers.size()) {
            ans.setText(arrayOfCheatedAnswers.get(cheatArrayIndex));
            cheatArrayIndex++;
        }
    }
    //Check if a word can be formed from the random letters
    private static boolean canFormWordFromLetters(String word, String letters) {
        // Convert the word and letters to lowercase
        word = word.toLowerCase();
        letters = letters.toLowerCase();

        //count char for letters
        int[] letterCount = new int[26];
        for (char c : letters.toCharArray()) {
            letterCount[c - 'a']++;
        }

        //can word be formed from the letters
        for (char c : word.toCharArray()) {
            if (letterCount[c - 'a'] <= 0) {
                return false;
            }
            letterCount[c - 'a']--;
        }
        return true;
    }
}

File: Client/GUI/GameGUI.py
import random
import tkinter as tk
from tkinter import messagebox

import threading

from . import RankGUI, ResultGUI, Homepage


class GameGUI:
    def __init__(self, wordy_game, username, room_name):
        self.wordy_game = wordy_game
        self.username = username
        self.room_name = room_name
        first_letter_line = ""
        second_letter_line = ""

        self.room_info = wordy_game.getGameRoomInformation(room_name, username)
        points = self.room_info[1]

        # Set up the GUI components
        self.root = tk.Tk()
        self.root.title("Wordy")

        # Set the size and location of the GUI
        self.root.geometry("800x650")
        self.root.resizable(False, False)
        self.root.wm_attributes("-topmost", 1)
        self.root.configure(bg='#1c2833')

        # to set the size of the window
        self.window_width = 800
        self.window_height = 650
        self.round_label = tk.Label(self.root, text="Round " + self.room_info[3], font=('Arial', 40), fg='#DEA54B', bg='#1c2833')
        self.round_label.pack(pady=20, padx=0)

        self.timer = tk.Label(self.root, text="Timer", font=('Arial', 30), fg='#DEA54B', bg='#1c2833')
        self.timer.pack(pady=30, padx=0)

        self.countdown_label = tk.Label(self.root, text=str(10), font=('Arial', 25), fg='#DBCBD8', bg='#1c2833')
        self.countdown_label.place(x=385, y=185)

        self.user_label = tk.Label(self.root, text=username, font=('Arial', 17), fg='#E8DAB2', bg='#1c2833')
        self.user_label.place(x=10, y=10)

        self.score_label = tk.Label(self.root, text="Score: " + str(points), font=('Arial', 17),  fg='#DEA54B', bg='#1c2833')
        self.score_label.place(x=self.window_width - 105, y=10)

        self.word_label = tk.Label(self.root, text=first_letter_line + "\n" + second_letter_line,
                                   font=('Orbitron', 40, 'bold'), justify='center', fg='#CE796B', bg='#1c2833')
        self.word_label.pack(pady=50, padx=0)

        random_letters = self.room_info[4]
        self.random_letters = random_letters
        self.shuffle_letters(random_letters)

        self.validation = tk.Label(self.root, text="", font=('Arial', 17), bg='#1c2833')
        self.validation.place(relx=0.5, rely=0.85, anchor="center")

        # Create answer entry and bind it to events
        self.answer_entry = tk.Entry(self.root, font=('Arial', 30), justify='center',fg='#1c2833', bg='#DEA54B')
        self.answer_entry.pack(pady=30, padx=0)
        self.answer_entry.bind('<Return>', self.validate_answer)
        self.answer_entry.bind("<KeyRelease>", lambda e: self.on_answer_changed(e))

        self.shuffle_button = tk.Button(self.root, text="SHUFFLE", font=('Arial', 11), command=self.click_shuffle, fg='#FFFFFC', bg='#8B9556')
        self.shuffle_button.place(x=548, y=440)

        # Start the countdown by calling the method after a delay
        self.root.after(1000, self.run_countdown)

        self.root.protocol("WM_DELETE_WINDOW", self.close_window)

        # Run the Tkinter event loop
        self.root.mainloop()

    def on_answer_changed(self, event):
        # Convert the entered text to uppercase and remove any digits
        text = self.answer_entry.get().upper()
        text = ''.join([i for i in text if not i.isdigit()])
        self.answer_entry.delete(0, tk.END)
        self.answer_entry.insert(0, text)

    previous_time_value = ""
    repetition_count = 0

    def run_countdown(self):
        room_info = self.wordy_game.getGameRoomInformation(self.room_name, self.username)
        countdown_value = int(room_info[0])

        # start of timer end error
        if room_info[0] == self.previous_time_value:
            self.repetition_count += 1
            if self.repetition_count == 7:
                self.root.destroy()
                messagebox.showerror("Error", "The game unexpectedly ended! Returning to main menu.")
                Homepage.Homepage(self.wordy_game, self.username)
        else:
            self.repetition_count = 1
        self.previous_time_value = room_info[0]
        # end of timer end error

        if countdown_value > 0:
            self.countdown_label.config(text=str(countdown_value))
            countdown_value -= 1
            self.root.after(1000, self.run_countdown)
        else:
            if not self.wordy_game.endRound(self.username, self.room_name):
                self.root.withdraw()
                RankGUI.RankGUI(self.wordy_game, self.username, self.room_name)
                self.root.destroy()
            else:
                self.root.withdraw()
                ResultGUI.ResultGUI(self.wordy_game, self.room_name, self.username)
                self.root.destroy()

    def click_shuffle(self):
        self.shuffle_letters(self.random_letters)

    def shuffle_letters(self, random_letters):
        letters_array = list(random_letters)

        # THE FISHER-YALES SHUFFLE ALGORITHM
        for i in range(len(letters_array) - 1, 0, -1):
            index = random.randint(0, i)
            temp = letters_array[index]
            letters_array[index] = letters_array[i]
            letters_array[i] = temp

        first_letter_line = ''.join(letters_array[:10]).strip()
        second_letter_line = ''.join(letters_array[10:]).strip()

        first_letter_line = ' '.join(list(first_letter_line))
        second_letter_line = ' '.join(list(second_letter_line))

        self.word_label.config(text=first_letter_line + "\n" + second_letter_line)

    def validate_answer(self, event=None):
        answer = self.answer_entry.get().strip()
        if answer:
            if self.wordy_game.submitWord(answer, self.username, int(self.room_info[3]), self.room_name):
                self.validation.config(text="ANSWER IS VALID", fg="green")

            else:
                self.validation.config(text="INVALID WORD FORMED", fg="red")
        else:
            self.validation.config(text="", fg="black")

        self.answer_entry.delete(0, tk.END)
        self.root.update()

    def close_window(self):
        self.wordy_game.logout(self.username)
        self.wordy_game.leaveGameRoom(self.username, self.room_name)
        self.root.destroy()
